File: J9Ch/src/J_7_Two_Pointers/Optional_11/_625_Partition_Array_II_Medium.py
# /** 625 Partition Array II


# 参考程序1
class Solution:
    # @param {int[]} nums an integer array
    # @param {int} low an integer
    # @param {int} high an integer
    # @return {int[]} any of the possible solutions
    def partition2(self, nums, low, high):
        # Write your code here
        if len(nums) <= 1:
            return

        pl, pr = 0, len(nums) - 1
        i = 0
        while i <= pr:
            if nums[i] < low:
                nums[pl], nums[i] = nums[i], nums[pl]
                pl += 1
                i += 1
            elif nums[i] > high:
                nums[pr], nums[i] = nums[i], nums[pr]
                pr -= 1
            else:
                i += 1

# 参考程序2
class Solution:
    # @param {int[]} nums an integer array
    # @param {int} low an integer
    # @param {int} high an integer
    # @return nothing
    def partition2(self, nums, low, high):
        # Write your code here
        left = 0
        right = len(nums) - 1

        # 首先把区间分为 < low 和 >= low 的两个部分
        while left <= right:
            while left <= right and nums[left] < low:
                left = left + 1
            while left <= right and nums[right] >= low:
                right = right - 1
            if left <= right:
                nums[right],nums[left] = nums[left],nums[right]
                left = left + 1
                right = right - 1

        right = len(nums) - 1
        # 然后从 >= low 的部分里分出 <= high 和 > high 的两个部分
        while left <= right:
            while left <= right and nums[left] <= high:
                left = left + 1
            while left <= right and nums[right] > high:
                right = right - 1
            if left <= right:
                nums[right],nums[left] = nums[left],nums[right]
                left = left + 1
                right = right - 1

File: J9Ch/src/J_7_Two_Pointers/Optional_11/_625_Partition_Array_II_Medium.cpp
/** 625 Partition Array II */


// 参考程序1
class Solution {
public:
    /**
     * @param nums an integer array
     * @param low an integer
     * @param high an integer
     * @return nothing
     */
    void partition2(vector<int>& nums, int low, int high) {
        // Write your code here
        if (nums.size() <= 1) {
            return;
        }

        int pl = 0, pr = nums.size() - 1;
        int i = 0;
        while (i <= pr) {
            if (nums[i] < low) {
                swap(nums, pl, i);
                pl++;
                i++;
            } else if (nums[i] > high) {
                swap(nums, pr, i);
                pr--;
            } else {
                i ++;
            }
        }
    }

    void swap(vector<int>& nums, int i, int j) {
        int tmp = nums[i];
        nums[i] = nums[j];
        nums[j] = tmp;
    }
};

//参考程序2
class Solution {
public:
    /**
     * @param nums an integer array
     * @param low an integer
     * @param high an integer
     * @return nothing
     */
    void partition2(vector<int>& nums, int low, int high) {
        // Write your code here
        int left = 0;
        int right = nums.size() - 1;

        // 首先把区间分为 < low 和 >= low 的两个部分
        while(left <= right) {
            while(left <= right && nums[left] < low) {
                left ++;
            }
            while(left <= right && nums[right] >= low) {
                right --;
            }

            if(left <= right) {
                int tmp = nums[left];
                nums[left] = nums[right];
                nums[right] = tmp;
                left ++;
                right --;
            }
        }

        // 然后从 >= low 的部分里分出 <= high 和 > high 的两个部分
        right = nums.size() - 1;
        while(left <= right) {
            while(left <= right && nums[left] <= high) {
                left ++;
            }
            while(left <= right && nums[right] > high) {
                right --;
            }
            if(left <= right) {
                int tmp = nums[left];
                nums[left] = nums[right];
                nums[right] = tmp;
                left ++;
                right --;
            }
        }
    }
};